pkg/install/stack/kubermatic-master/nginxingress.go

```diff
@@ -0,0 +1,223 @@
+
+	chart, err := helm.LoadChart(filepath.Join(opt.ChartsDirectory, "nginx-ingress-controller"))
+	if err != nil {
+		return fmt.Errorf("failed to load Helm chart: %v", err)
+	}
+
+	if err := util.EnsureNamespace(ctx, sublogger, kubeClient, NginxIngressControllerNamespace); err != nil {
+		return fmt.Errorf("failed to create namespace: %v", err)
+	}
+
+	release, err := util.CheckHelmRelease(ctx, sublogger, helmClient, NginxIngressControllerNamespace, NginxIngressControllerReleaseName)
+	if err != nil {
+		return fmt.Errorf("failed to check to Helm release: %v", err)
+	}
+
+	// if version older than 1.3.0 is installed, we must perform a migration
+	// by deleting the old deployment object for the controller
+	v13 := semver.MustParse("1.3.0")
+	backupTS := time.Now().Format("2006-01-02T150405")
+
+	if release != nil && release.Version.LessThan(v13) && !chart.Version.LessThan(v13) {
+		if !opt.EnableNginxIngressMigration {
+			sublogger.Warn("To upgrade nginx-ingress-controller to a new version, installer")
+			sublogger.Warn("will remove the old deployment object before proceeding with the upgrade.")
+			sublogger.Warn("Rerun the installer with --migrate-nginx-ingress to enable the migration process.")
+			sublogger.Warn("Please refer to the KKP 2.19 upgrade notes for more information.")
+
+			return fmt.Errorf("user must acknowledge the migration using --migrate-nginx-ingress")
+		}
+
+		err = upgradeNginxIngress(ctx, sublogger, kubeClient, helmClient, opt, chart, release, backupTS)
+		if err != nil {
+			return fmt.Errorf("failed to upgrade nginx-ingress-controller: %v", err)
+		}
+
+	}
+
+	// do not perform an atomic installation, as this will make Helm wait for the LoadBalancer to
+	// get an IP and this can require manual intervention based on the target environment
+	sublogger.Info("Deploying Helm chart...")
+	if err := util.DeployHelmChart(ctx, sublogger, helmClient, chart, NginxIngressControllerNamespace, NginxIngressControllerReleaseName, opt.HelmValues, false, opt.ForceHelmReleaseUpgrade, release); err != nil {
+		return fmt.Errorf("failed to deploy Helm release: %v", err)
+	}
+
+	if err := waitForNginxIngressWebhook(ctx, sublogger, kubeClient, helmClient, opt); err != nil {
+		return fmt.Errorf("failed to verify that the webhook is functioning: %v", err)
+	}
+
+	logger.Info("✅ Success.")
+
+	return nil
+}
+
+func upgradeNginxIngress(
+	ctx context.Context,
+	logger *logrus.Entry,
+	kubeClient ctrlruntimeclient.Client,
+	helmClient helm.Client,
+	opt stack.DeployOptions,
+	chart *helm.Chart,
+	release *helm.Release,
+	backupTS string,
+) error {
+	logger.Infof("%s: %s detected, performing upgrade to %s…", release.Name, release.Version.String(), chart.Version.String())
+	// 1: find the old deployment
+	logger.Info("Backing up old ingress deployment...")
+	deploymentsList := &unstructured.UnstructuredList{}
+	deploymentsList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "apps",
+		Kind:    "DeploymentList",
+		Version: "v1",
+	})
```

Comment:
Why not use [`appsv1.DeploymentList`](https://pkg.go.dev/k8s.io/api@v0.22.4/apps/v1#DeploymentList) here?

Reply:
because as far as I understand, our `util.DumpResources` expects an `[]Unstructured` and I didn't want to convert the other way around, but might be wrong on that.

Reply:
Ah, yeah that seems right. Never mind then :)